plat/kvm/arm: Replace psci functions with generic SMCCC calls

razvanvirtan opened the pull request:
### Prerequisite checklist
 - [X] Read the [contribution guidelines](https://docs.unikraft.org/contribute.html) regarding submitting new changes to the project;
 - [X] Tested your changes against relevant architectures and platforms;
 - [X] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target
 - Architecture(s): [ARM, ARM64]
 - Platform(s): [KVM]


### Additional configuration


### Description of changes
Until now, the kvm platform setup used hvc and smc wrappers that were PSCI specific. After the generic SMCCC interface was
introduced, there is no need to do this anymore.
This commit removes the `psci_arm64.S` and replaces the functions implemented there with generic `smccc_smc()`
and `smccc_hvc()` calls.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

razvand commented:
@razvanvirtan , @michpappas, @rene , I marked this PR for release 0.9.

michpappas commented on plat/common/include/arm/arm64/cpu.h:
```diff
@@ -136,13 +137,10 @@ static inline void _init_cpufeatures(void)
-extern smcc_psci_callfn_t smcc_psci_call;
-int32_t smcc_psci_hvc_call(uint32_t, uint64_t, uint64_t, uint64_t);
-int32_t smcc_psci_smc_call(uint32_t, uint64_t, uint64_t, uint64_t);
+typedef void (*smccc_callfn_t)(struct smccc_args *args);
```
Hi @razvanvirtan looks good, thanks a lot 👍🏼  Just one comment from me on here: We already have a similar requirement to declare a pointer to function in `smccc.c`. Perhaps it would be nice to declare a common type in `smccc.h`, and use it in both `smccc.c` and for psci too.
```
typedef void (*smccc_conduit_fn_t)(struct smccc_args *args);
```
Btw, I choose `smccc_conduit_fn_t` here instead of `smccc_callfn_t`, as Arm uses the term "conduit" in their documents to refer to the method used.

michpappas commented on plat/common/include/arm/arm64/cpu.h:
```diff
@@ -136,13 +137,10 @@ static inline void _init_cpufeatures(void)
 /*
- * PSCI conduit method to call functions, based on the SMC Calling
- * Convention.
+ * SMCCC conduit method to call functions, based on the SMC Calling Convention.
```
Actually this comment was correct. Remember that we can use a different conduit for PSCI than, say, for accessing a Secure OS. That's pretty common when PSCI calls are emulated by the HV, which could require using `hvc`.

rene commented on plat/common/arm/cpu_native.c:
```diff
@@ -37,6 +37,8 @@
 #include <arm/psci.h>
```
It should be supported on ARM32 now, right?

razvanvirtan replied:
Yes, I missed removing those guards, my bad.

michpappas commented:
@rene same here. If you have any comments please provide your update, otherwise we need your approval to merge. Thanks!

rene reviewed (approved):
@razvanvirtan , I have no more comments, looks good to me.

michpappas commented:
@rene, please post an RB comment so that the bot adds your approval it to the commit message, thanks 🙂

rene commented:
Reviewed-by: Renê de Souza Pinto <rene@renesp.com.br>

michpappas reviewed (changes requested):
Hi @razvan just these minor ones from me

michpappas reviewed (changes requested):
Hi @razvanvirtan, looks good. Would it be possible if you split this into two separate commits, ie

- One commit that introduces the new type in `smccc.h` and updates `smccc.c` to use it (if you also rename `_conduit` to `conduit` it would be great).
- One commit with the rest of your changes.

From the functionality perspective it looks good to go, so once you update I can trigger the merge.

razvanvirtan committed:
plat/common/arm: Introduce new smccc conduit function type

This commit introduces a new data type for the smccc conduit
functions.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

razvanvirtan committed:
plat/kvm/arm: Replace psci functions with generic SMCCC calls

Until now, the kvm platform setup used hvc and smc wrappers
that were PSCI specific. After the generic SMCCC interface was
introduced, there is no need to do this anymore.
This commit removes the `psci_arm64.S` and replaces
the functions implemented there with generic `smccc_smc()`
and `smccc_hvc()` calls.

Signed-off-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

michpappas commented:
@razvanvirtan thanks a lot!

Approved-by: Michalis Pappas <michalis.pappas@opensynergy.com>
